pkg/glusterfs/controllerserver.go

```diff
@@ -405,15 +454,243 @@ func (cs *ControllerServer) ControllerGetCapabilities(ctx context.Context, req *
+		if parseErr != nil {
+			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		}
+		startToken = int32(i)
+	}
+
+	if len(req.GetSnapshotId()) != 0 {
+		return cs.listSnapshotFromID(req.GetSnapshotId())
+	}
+
+	//If volume id is sent
+	if len(req.GetSourceVolumeId()) != 0 {
+		snaplist, err = cs.client.SnapshotList(req.SourceVolumeId)
+		if err != nil {
+			errResp := cs.client.LastErrorResponse()
+			if errResp != nil && errResp.StatusCode == http.StatusNotFound {
+				resp := csi.ListSnapshotsResponse{}
+				return &resp, nil
+			}
+			glog.Errorf("failed to list snapshots %v", err)
+			return nil, status.Errorf(codes.Internal, "ListSnapshot - failed to get snapshots %s", err.Error())
+		}
+	} else {
+		//Get all snapshots
+		snaplist, err = cs.client.SnapshotList("")
+		if err != nil {
+			glog.Errorf("failed to list snapshots %v", err)
+			return nil, status.Errorf(codes.Internal, "failed to get snapshots %s", err.Error())
+		}
+	}
+
+	return cs.doPagination(req, snaplist, startToken)
+}
+
+func (cs *ControllerServer) listSnapshotFromID(snapID string) (*csi.ListSnapshotsResponse, error) {
+	var entries []*csi.ListSnapshotsResponse_Entry
+	snap, err := cs.GfDriver.client.SnapshotInfo(snapID)
+	if err != nil {
+		errResp := cs.client.LastErrorResponse()
+		if errResp != nil && errResp.StatusCode == http.StatusNotFound {
+			resp := csi.ListSnapshotsResponse{}
+			return &resp, nil
+		}
+		glog.Errorf("failed to get snapshot info %v", err)
+		return nil, status.Errorf(codes.NotFound, "ListSnapshot - failed to get snapshot info %s", err.Error())
+
+	}
+	entries = append(entries, &csi.ListSnapshotsResponse_Entry{
+		Snapshot: &csi.Snapshot{
+			Id:             snap.VolInfo.Name,
+			SourceVolumeId: snap.ParentVolName,
+			CreatedAt:      snap.CreatedAt.Unix(),
+			SizeBytes:      (int64(snap.VolInfo.Capacity)) * utils.MB,
+			Status: &csi.SnapshotStatus{
+				Type: csi.SnapshotStatus_READY,
+			},
+		},
+	})
+
+	resp := csi.ListSnapshotsResponse{}
+	resp.Entries = entries
+	return &resp, nil
+
+}
+func (cs *ControllerServer) doPagination(req *csi.ListSnapshotsRequest, snapList api.SnapListResp, startToken int32) (*csi.ListSnapshotsResponse, error) {
+	if req.GetStartingToken() != "" && int(startToken) > len(snapList) {
+		return nil, status.Error(codes.Aborted, "invalid starting token")
+	}
+
+	var entries []*csi.ListSnapshotsResponse_Entry
+	for _, snap := range snapList {
+		for _, s := range snap.SnapList {
+			entries = append(entries, &csi.ListSnapshotsResponse_Entry{
+				Snapshot: &csi.Snapshot{
+					Id:             s.VolInfo.Name,
+					SourceVolumeId: snap.ParentName,
+					CreatedAt:      s.CreatedAt.Unix(),
+					SizeBytes:      (int64(s.VolInfo.Capacity)) * utils.MB,
+					Status: &csi.SnapshotStatus{
+						Type: csi.SnapshotStatus_READY,
+					},
+				},
+			})
+		}
+
+	}
+
+	//TODO need to remove paginating code once glusterd2 supports pagination
```

Comment:
lets open an issue and mention it here.

Reply:
already have an issue in gd2 https://github.com/gluster/glusterd2/issues/372

Reply:
@Madhu-1 what I meant was mention the issue in the source code comments.